### ISSUE TITLE
dev_cli.sh fails to launch docker in rootless mode

### ISSUE THREAD

**Issue by waywardgeek** [cloud-hypervisor/cloud-hypervisor#5411]
**Describe the bug**
TL;DR: dev_cli.sh fails to launch Docker in rootless mode, because the group and user ID should be 0, not the user's group and user ID.

Should I submit a PR adding a flag for rootless mode, that would set uid and gid to 0?  Or maybe check an environment variable?

The details...

Running 
```
$ ./scripts/dev_cli.sh build --release
```
in Docker rootless mode results in:
```
cloud-hypervisor$ ./scripts/dev_cli.sh build --release
+ CLI_NAME='Cloud Hypervisor'
+ CTR_IMAGE_TAG=ghcr.io/cloud-hypervisor/cloud-hypervisor
+ CTR_IMAGE_VERSION=20230316-0
+ CTR_IMAGE=ghcr.io/cloud-hypervisor/cloud-hypervisor:20230316-0
+ DOCKER_RUNTIME=docker
+++ dirname ./scripts/dev_cli.sh
++ cd ./scripts
++ pwd
+ CLH_SCRIPTS_DIR=/home/waywardgeek/cloud-hypervisor/scripts
++ cd /home/waywardgeek/cloud-hypervisor/scripts/..
++ pwd
+ CLH_ROOT_DIR=/home/waywardgeek/cloud-hypervisor
+ CLH_BUILD_DIR=/home/waywardgeek/cloud-hypervisor/build
+ CLH_CARGO_TARGET=/home/waywardgeek/cloud-hypervisor/build/cargo_target
+ CLH_DOCKERFILE=/home/waywardgeek/cloud-hypervisor/scripts/../resources/Dockerfile
+ CLH_CTR_BUILD_DIR=/tmp/cloud-hypervisor/ctr-build
+ CLH_INTEGRATION_WORKLOADS=/home/waywardgeek/workloads
+ CTR_CLH_ROOT_DIR=/cloud-hypervisor
+ CTR_CLH_CARGO_BUILT_DIR=/cloud-hypervisor/build
+ CTR_CLH_CARGO_TARGET=/cloud-hypervisor/build/cargo_target
+ CTR_CLH_INTEGRATION_WORKLOADS=/root/workloads
+ CTR_CLH_NET=bridge
+ CARGO_REGISTRY_DIR=/home/waywardgeek/cloud-hypervisor/build/cargo_registry
+ CARGO_GIT_REGISTRY_DIR=/home/waywardgeek/cloud-hypervisor/build/cargo_git_registry
+ CARGO_TARGET_DIR=/home/waywardgeek/cloud-hypervisor/build/cargo_target
+ '[' 2 = 0 ']'
+ '[' 2 -gt 0 ']'
+ case "$1" in
+ break
+ declare -f cmd_build
+ ok_or_die 'Unknown command: build. Please use `./scripts/dev_cli.sh help` for help.'
+ code=0
+ [[ 0 -eq 0 ]]
+ cmd=cmd_build
+ shift
+ cmd_build --release
+ build=debug
+ libc=gnu
+ hypervisor=kvm
+ features_build=
+ exported_device=/dev/kvm
+ '[' 1 -gt 0 ']'
+ case "$1" in
+ build=release
+ shift
+ '[' 0 -gt 0 ']'
+ ensure_build_dir
+ for dir in "$CLH_BUILD_DIR" "$CLH_INTEGRATION_WORKLOADS" "$CLH_CTR_BUILD_DIR" "$CARGO_TARGET_DIR" "$CARGO_REGISTRY_DIR" "$CARGO_GIT_REGISTRY_DIR"
+ mkdir -p /home/waywardgeek/cloud-hypervisor/build
+ '[' -x /home/waywardgeek/cloud-hypervisor/build ']'
+ '[' -w /home/waywardgeek/cloud-hypervisor/build ']'
+ for dir in "$CLH_BUILD_DIR" "$CLH_INTEGRATION_WORKLOADS" "$CLH_CTR_BUILD_DIR" "$CARGO_TARGET_DIR" "$CARGO_REGISTRY_DIR" "$CARGO_GIT_REGISTRY_DIR"
+ mkdir -p /home/waywardgeek/workloads
+ '[' -x /home/waywardgeek/workloads ']'
+ '[' -w /home/waywardgeek/workloads ']'
+ for dir in "$CLH_BUILD_DIR" "$CLH_INTEGRATION_WORKLOADS" "$CLH_CTR_BUILD_DIR" "$CARGO_TARGET_DIR" "$CARGO_REGISTRY_DIR" "$CARGO_GIT_REGISTRY_DIR"
+ mkdir -p /tmp/cloud-hypervisor/ctr-build
+ '[' -x /tmp/cloud-hypervisor/ctr-build ']'
+ '[' -w /tmp/cloud-hypervisor/ctr-build ']'
+ for dir in "$CLH_BUILD_DIR" "$CLH_INTEGRATION_WORKLOADS" "$CLH_CTR_BUILD_DIR" "$CARGO_TARGET_DIR" "$CARGO_REGISTRY_DIR" "$CARGO_GIT_REGISTRY_DIR"
+ mkdir -p /home/waywardgeek/cloud-hypervisor/build/cargo_target
+ '[' -x /home/waywardgeek/cloud-hypervisor/build/cargo_target ']'
+ '[' -w /home/waywardgeek/cloud-hypervisor/build/cargo_target ']'
+ for dir in "$CLH_BUILD_DIR" "$CLH_INTEGRATION_WORKLOADS" "$CLH_CTR_BUILD_DIR" "$CARGO_TARGET_DIR" "$CARGO_REGISTRY_DIR" "$CARGO_GIT_REGISTRY_DIR"
+ mkdir -p /home/waywardgeek/cloud-hypervisor/build/cargo_registry
+ '[' -x /home/waywardgeek/cloud-hypervisor/build/cargo_registry ']'
+ '[' -w /home/waywardgeek/cloud-hypervisor/build/cargo_registry ']'
+ for dir in "$CLH_BUILD_DIR" "$CLH_INTEGRATION_WORKLOADS" "$CLH_CTR_BUILD_DIR" "$CARGO_TARGET_DIR" "$CARGO_REGISTRY_DIR" "$CARGO_GIT_REGISTRY_DIR"
+ mkdir -p /home/waywardgeek/cloud-hypervisor/build/cargo_git_registry
+ '[' -x /home/waywardgeek/cloud-hypervisor/build/cargo_git_registry ']'
+ '[' -w /home/waywardgeek/cloud-hypervisor/build/cargo_git_registry ']'
+ ensure_latest_ctr
+ '[' 20230316-0 = local ']'
+ docker pull ghcr.io/cloud-hypervisor/cloud-hypervisor:20230316-0
20230316-0: Pulling from cloud-hypervisor/cloud-hypervisor
Digest: sha256:3accf6ac8436472c0c80a3f368bd0b0ca214c3856d6efa103c50c1248afbf756
Status: Image is up to date for ghcr.io/cloud-hypervisor/cloud-hypervisor:20230316-0
ghcr.io/cloud-hypervisor/cloud-hypervisor:20230316-0
+ '[' 0 -ne 0 ']'
+ ok_or_die 'Error pulling/building container image. Aborting.'
+ code=0
+ [[ 0 -eq 0 ]]
+ process_volumes_args
+ '[' -z '' ']'
+ return
+ [[ kvm = \k\v\m ]]
+ [[ kvm = \m\s\h\v ]]
++ uname -m
+ target=x86_64-unknown-linux-gnu
+ cargo_args=("$@")
+ '[' release = release ']'
+ cargo_args+=("--release")
+ cargo_args+=(--target "$target")
+ rustflags=
+ target_cc=
++ uname -m
+ '[' x86_64 = aarch64 ']'
++ id -u
++ id -g
+ docker run --user 292356:89939 --workdir /cloud-hypervisor --rm --volume /dev/kvm --volume /home/waywardgeek/cloud-hypervisor:/cloud-hypervisor --env RUSTFLAGS= --env TARGET_CC= ghcr.io/cloud-hypervisor/cloud-hypervisor:20230316-0 cargo build --all --target-dir /cloud-hypervisor/build/cargo_target --release --target x86_64-unknown-linux-gnu
docker: Error response from daemon: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: unable to setup user: setgroups: invalid argument: unknown.
```
With a small change to dev_cli.sh, I'm able to use dev_cli.sh for building and testing:

```
loud-hypervisor$ git diff scripts/dev_cli.sh
diff --git a/scripts/dev_cli.sh b/scripts/dev_cli.sh
index 389f460e..6be120db 100755
--- a/scripts/dev_cli.sh
+++ b/scripts/dev_cli.sh
@@ -4,6 +4,8 @@
 # Copyright © 2020 Intel Corporation
 # SPDX-License-Identifier: Apache-2.0
 
+set -x
+
 CLI_NAME="Cloud Hypervisor"
 
 CTR_IMAGE_TAG="ghcr.io/cloud-hypervisor/cloud-hypervisor"
@@ -145,7 +147,7 @@ fix_dir_perms() {
         --volume /dev:/dev \
         --volume "$CLH_ROOT_DIR:$CTR_CLH_ROOT_DIR" $exported_volumes \
         "$CTR_IMAGE" \
-        chown -R "$(id -u):$(id -g)" "$CTR_CLH_ROOT_DIR"
+        chown -R "0:0" "$CTR_CLH_ROOT_DIR"
 
     return "$1"
 }
@@ -290,7 +292,7 @@ cmd_build() {
     fi
 
     $DOCKER_RUNTIME run \
-        --user "$(id -u):$(id -g)" \
+        --user "0:0" \
         --workdir "$CTR_CLH_ROOT_DIR" \
         --rm \
         --volume $exported_device \
@@ -310,7 +312,7 @@ cmd_clean() {
     ensure_latest_ctr
 
     $DOCKER_RUNTIME run \
-        --user "$(id -u):$(id -g)" \
+        --user "0:0" \
         --workdir "$CTR_CLH_ROOT_DIR" \
         --rm \
         --volume "$CLH_ROOT_DIR:$CTR_CLH_ROOT_DIR" $exported_volumes \
(END)

```

**To Reproduce**
1) Enable rootless mode by [following these instructions over at Project-Oak](https://github.com/project-oak/oak/blob/main/docs/development.md#rootless-docker).
2) Follow the CHV [Clone and Build instructions](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/docs/building.md#building-cloud-hypervisor).

The first failure I ran into was when executing `$ ./scripts/dev_cli.sh build --release`.

**Version**

cloud-hypervisor v31.0-45-g0d58e51a-dirty

I built from source following the CHV [BuildingCloud Hypervisor](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/docs/building.md#building-cloud-hypervisor) almost exactly.  I already had rustup installed, so I skipped the curl step, and instead ran `rustup update` and continued with the remaining instructions.

The build command was:
```
$ cargo build --release --target=x86_64-unknown-linux-musl --all
```
The build was successful.

**VM configuration**

What command line did you run (or JSON config data):
The dev_cli.sh script executed:
```
docker run --user 292356:89939 --workdir /cloud-hypervisor --rm --volume /dev/kvm --volume /home/waywardgeek/cloud-hypervisor:/cloud-hypervisor --env RUSTFLAGS= --env TARGET_CC= ghcr.io/cloud-hypervisor/cloud-hypervisor:20230316-0 cargo build --all --target-dir /cloud-hypervisor/build/cargo_target --release --target x86_64-unknown-linux-gnu
```
Guest OS version details:
The guest OS is included in the Docker container CHV uses with dev_cli.sh.  I am not familiar with its setup.

Host OS version details:
The host OS is : gLinux, a Google internal OS based on Debian.

**Logs**

The failure is encountered launching the Docker container.  CHV is never run.

Linux kernel output:

**Comment by likebreath** [comment]
This sounds fine to me. Thank you for the contributions.
